pjsip/src/pjsip-ua/sip_inv.c

```diff
@@ -181,6 +188,69 @@ struct tsx_inv_data
     pj_bool_t            has_sdp;   /* Message with SDP?                    */
 };
 
+
+/* Add Reason header to request (rfc3326) or Warning header to error response
+ * (rfc3261 20.43).
+ */
+static pj_status_t add_reason_warning_hdr(pjsip_tx_data *tdata,
+                                          unsigned code,
+                                          const char *str)
+{
+    pj_str_t reason;
+    pj_cstr(&reason, str);
+    if (tdata->msg->type == PJSIP_REQUEST_MSG) {
+
+        /* Add Reason header in request messages */
+        pjsip_hdr *hdr;
+        const pj_str_t hname = { "Reason", 6 };
+        pj_str_t hvalue;
+
+        PJ_ASSERT_RETURN(code < 1000, PJ_EINVAL);
+
+        hvalue.ptr = (char*)pj_pool_alloc(tdata->pool,
+                                          3 +              /* "SIP" */
+                                          11 +             /* " ;cause=code */
+                                          reason.slen+10); /* " ;text=".." */
+
+        if (!hvalue.ptr)
+            return PJ_ENOMEM;
+
+        hvalue.slen = pj_ansi_sprintf(hvalue.ptr,
+                                      "SIP ;cause=%u ;text=\"%.*s\"",
+                                      code, (int)reason.slen, reason.ptr);
+        hdr = (pjsip_hdr*)
+               pjsip_generic_string_hdr_create(tdata->pool, &hname, &hvalue);
+        if (hdr)
+            pjsip_msg_add_hdr(tdata->msg, (pjsip_hdr*)hdr);
+        else
+            return PJ_ENOMEM;
+
+    } else {
+
+        /* Add Warning header in response messages */
+        pjsip_warning_hdr *hdr;
+        pjsip_dialog *dlg = pjsip_tdata_get_dlg(tdata);
+
+        PJ_ASSERT_RETURN(dlg, PJ_EINVAL);
+
+        // Let's always use code 390 for now.
+        //PJ_ASSERT_RETURN(code >= 300 && code < 400, PJ_EINVAL);
+        code = 390;
```

Comment:
Since code and warn-code are actually different, perhaps better to change this to:
```
        // Let's always use code 390 for now.
        // TODO: unsigned warn_code = map_code_to_warn_code(code);
        unsigned warn_code = 390;
```